src/rust/engine/src/context.rs

```diff
@@ -35,12 +35,16 @@ impl Core {
     types: Types,
     build_root: PathBuf,
     ignore_patterns: Vec<String>,
+    work_dir: PathBuf,
   ) -> Core {
+    let mut snapshots_dir = work_dir.clone();
```

Comment:
Rather than cloning the path here in order to make it mutable, you can mark the parameter `mut`, and then directly mutate it.
```
fn new(
  ..,
  mut work_dir: PathBuf,
) -> Core {
  work_dir.push("snapshots");
```

This might seem strange, since in pass-by-reference-value languages that would mean mutating the variable in a way that the caller's might notice. But this parameter is coming in by value, so the caller cannot be affected: they gave ownership to this function (ie, the value "moved into" the function).